core/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -87,9 +83,8 @@ public GatewayAllocator(Settings settings, TransportNodesListGatewayStartedShard
         logger.debug("using initial_shards [{}]", initialShards);
     }
 
-    public void setReallocation(final ClusterService clusterService, final AllocationService allocationService) {
-        this.clusterService = clusterService;
-        this.allocationService = allocationService;
+    public void setReallocation(final ClusterService clusterService, final RoutingService routingService) {
```

Comment:
the cycling problem will still remain, do you mean fold RoutingService into ClusterService? I kinda like that ClusterService has no notion or knowledge of allocation / routing, and just deals with cluster state in a generic form